src/utils.c

```diff
@@ -150,6 +151,143 @@ char *string_tuple_extract_head(char *tuple, int result_buf_size, uint8_t *resul
     return tuple + i + (int)(tuple[i] == ',');
 }
 
+hyperloglog_t *hll_rle_decode(const char* hll_rle_str)
+{
+    enum State {
+        LEN,
+        LEN_E,
+        M_VAL
+    };
+    uint8_t p = hex_byte_to_byte(&hll_rle_str[0]);
+    enum State curstate = LEN;
+    hyperloglog_t *res = hll_init(p);
+    uint32_t len = 0;
+    uint32_t register_index = 0;
+    hll_rle_str += 4;
+    /* if we get passed a json string it ends with a qutation mark
+       handle both just in case 
+    */
+    while((*hll_rle_str != '\0') && (*hll_rle_str != '\"')){
+        uint32_t val = hex_byte_to_byte(hll_rle_str);
+        if(curstate == LEN){
+            len = val & ((1 <<7) -1);
+            // If MSB is 1 length didn't fit in 1 byte
+            if (val & (1<<7)){
+                curstate = LEN_E;
+            } else {
+                curstate = M_VAL;
+            }
+        } else if (curstate == LEN_E){
+            len |= (val << 7);
+            curstate = M_VAL; 
+        }
+        else if (curstate == M_VAL){
+            uint32_t stop = register_index + len;
+            for(; register_index < stop; register_index++){
+                res->M[register_index] = val; 
+            }
+            len =0;
+            curstate = LEN;
+        } else {
+            return NULL;
+        }
+        hll_rle_str += 2;
+    }
+    return res;
+}
+
+uint8_t hex_byte_to_byte(const char* h){
+    uint8_t res = 0;
+    for(int i=0; i<2; i++){
+        uint8_t byte = h[i];
+        // This shouldn't be a problem because we should always get even strings
+        // but just in case. Segfaults are a bitch
+        if(byte != 0){
+            if ( h[i]>= '0' && h[i] <= '9') byte = byte - '0';
+            else if (byte >= 'a' && byte <='f') byte = byte - 'a' + 10;
+            else if (byte >= 'A' && byte <='F') byte = byte - 'A' + 10;
+        }
+        res = (res << 4) | (byte & 0xF);
+    }
+    return res;
+}
+
+char *run_length_encode(char *in, size_t size, int *out_size) {
+    char *out;
+    *out_size = 0;
+    if (size == 0) {
+        out = malloc(1);
+        *out = '\0';
+        return out;
+    } else {
+        // Allocate twice the size of the input string
+        // which is the max amount memory we could need,
+        // in the worst case.
```

Comment:
Nitpick but comments style in this file is not consistent with itself and the rest of the code.